common/src/main/java/org/apache/sedona/common/raster/Rasterization.java

```diff
@@ -0,0 +1,598 @@
+
+    for (int i = 0; i < geom.getNumGeometries(); i++) {
+      Geometry subGeom = geom.getGeometryN(i);
+      Envelope2D subGeomExtent = rasterizeGeomExtent(subGeom, raster, metadata, allTouched);
+      rasterizeGeometry(raster, metadata, subGeom, params, subGeomExtent, value, allTouched);
+    }
+  }
+
+  private static void rasterizePoint(
+      Geometry geom, RasterizationParams params, Envelope2D geomExtent, double value) {
+
+    int startX = (int) Math.round(((geomExtent.getMinX() - params.upperLeftX) / params.scaleX));
+    int startY = (int) Math.round(((geomExtent.getMinY() - params.upperLeftY) / params.scaleY));
+    int x = startX;
+    int y = startY;
+
+    for (double worldY = geomExtent.getMinY();
+        worldY < geomExtent.getMaxY();
+        worldY += params.scaleY, y++) {
+      x = startX;
+      for (double worldX = geomExtent.getMinX();
+          worldX < geomExtent.getMaxX();
+          worldX += params.scaleX, x++) {
+
+        // Flip y-axis (since raster Y starts from top-left)
+        int yIndex = -y - 1;
+
+        // Create envelope for this pixel
+        double cellMaxX = worldX + params.scaleX;
+        double cellMaxY = worldY + params.scaleY;
+
+        boolean intersects =
+            geom.intersects(JTS.toGeometry(new Envelope(worldX, cellMaxX, worldY, cellMaxY)));
+
+        if (intersects) {
+          params.writableRaster.setSample(x, yIndex, 0, value);
+        }
+      }
+    }
+  }
+
+  private static void rasterizeLineString(
+      Geometry geom, RasterizationParams params, double value, Envelope2D geomExtent) {
+    for (int i = 0; i < geom.getNumGeometries(); i++) {
+      LineString line = (LineString) geom.getGeometryN(i);
+      Coordinate[] coords = line.getCoordinates();
+
+      for (int j = 0; j < coords.length - 1; j++) {
+        // Extract start and end points for the segment
+        LineSegment clippedSegment =
+            clipSegmentToRasterBounds(coords[j], coords[j + 1], geomExtent);
+
+        Coordinate start;
+        Coordinate end;
+        if (clippedSegment != null) {
+          start = new Coordinate(clippedSegment.p0.x, clippedSegment.p0.y);
+          end = new Coordinate(clippedSegment.p1.x, clippedSegment.p1.y);
+        } else {
+          continue; // Skip case where segment is completely outside geomExtent
+        }
+
+        double x0 = (start.x - params.upperLeftX) / params.scaleX;
+        double y0 = (params.upperLeftY - start.y) / params.scaleY;
+        double x1 = (end.x - params.upperLeftX) / params.scaleX;
+        double y1 = (params.upperLeftY - end.y) / params.scaleY;
+
+        // Apply Bresenham for this segment
+        drawLineBresenham(params, x0, y0, x1, y1, value, 0.2);
+      }
+    }
+  }
+
+  // Modified Bresenham with Fractional Steps
+  private static void drawLineBresenham(
+      RasterizationParams params,
+      double x0,
+      double y0,
+      double x1,
+      double y1,
+      double value,
+      double stepSize) {
```

Comment:
Why not use a traditional Bresenham's line algorithm using integer arithmetics? Using a `stepSize = 0.2` here introduces lots of repeated setSample operations.

Reply:
I found that GDAL uses a [more complex approach](https://github.com/OSGeo/gdal/blob/v3.10.2/alg/llrasterize.cpp#L651-L710) for allTouched mode: it figures out which step to take (right or up/down) to step into the next pixel. This could be more reliable than using a fractional step of 0.2.

I am OK with the current approach, but I think we can switch to GDAL's approach in the future.